### PR TITLE
update GD32F30x firmware library to 2.1.4

### PR DESCRIPTION
**Pull request by tlyu**
Vendor release 2.1.4 dated 2022-07-11. URLs:
https://www.gd32mcu.com/download/down/document_id/129/path_type/1
https://www.gd32mcu.com/data/documents/toolSoftware/GD32F30x_Firmware_Library_V2.1.4.rar

Notable USBD library improvements:

* Track control transfer state, avoiding some race conditions that
  caused looping ping-pong status transactions.

* Change some API types from 8-bit to 16-bit, avoiding truncation.

* Handle configuration descriptor sets longer than 255 bytes.

This PR also re-applies some local patches, and undoes an incorrect vendor change to the USBD library.

### CODE REVIEW

File: system/GD32F30x_firmware/GD32F30x_standard_peripheral/Include/gd32f30x_ctc.h [hunk 1]
```diff
@@ -97,7 +97,6 @@ OF SUCH DAMAGE.
 #define CTL1_REFSEL(regval)                              (BITS(28,29) & ((uint32_t)(regval) << 28))
 #define CTC_REFSOURCE_GPIO                               CTL1_REFSEL(0)               /*!< GPIO is selected */
 #define CTC_REFSOURCE_LXTAL                              CTL1_REFSEL(1)               /*!< LXTAL is selected */
-#define CTC_REFSOURCE_USB_SOF                            CTL1_REFSEL(2)               /*!< USBD_SOF or USBFS_SOF is selected */
```
Comment (obra) [inline system/GD32F30x_firmware/GD32F30x_standard_peripheral/Include/gd32f30x_ctc.h]: Would it be possible for us to locally revert the change that breaks the SOF CTC source?

Comment (tlyu) [reply system/GD32F30x_firmware/GD32F30x_standard_peripheral/Include/gd32f30x_ctc.h]: Yeah, we could do that. Do you want that added to this PR or a separate one?

I know it appears to work, at least on the one piece of hardware that I have access to. Unfortunately, we still don't know why it was silently removed. We might not want to use it if it causes some non-obvious catastrophic behavior some of the time.

**Comment by obra** [comment]
I have devices in the field that need it, but wouldn't depend on it for
anything new.

I'd rather it go in in the same PR, but in a standalone commit.
